### PR TITLE
Don't send updates to sidecars for Gateway updates

### PR DESCRIPTION
**Pull request by howardjohn**
This change expands the proxyNeedsPush logic to allow determining if a
push is required based on the config type. This sets the groundwork for
future improvements. Additionally, it adds the first check using this
functionality, limiting updates for Gateways to only router types.

This copies a lot of code from https://github.com/istio/istio/pull/16076/files

A completely synthetic benchmark of applying a gateway a lot: https://snapshot.raintank.io/dashboard/snapshot/ZhH1bDnp8X0BT9ZtmlExSDR5QMmR9GvX. This obviously is not realistic since no one is applying gateways this much, but does show it works

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
I agree this is a good start.
But what's more, i think we need an overall design to make it clear the relationship between   
rules and sidecars, also rules and xDS.

File: pilot/pkg/proxy/envoy/v2/ads.go [hunk 1]
```diff
@@ -702,25 +706,45 @@ func (s *DiscoveryServer) startPush(req *model.PushRequest) {
...
 
-func proxyNeedsPush(con *XdsConnection, targetNamespaces map[string]struct{}) bool {
+func ProxyNeedsPush(proxy *model.Proxy, targetNamespaces map[string]struct{}, configs map[string]struct{}) bool {
 	if !features.ScopePushes.Get() {
 		// If push scoping is not enabled, we push for all proxies
 		return true
 	}
 
+	// appliesToProxy starts as false, we will set it to true if we encounter any configs that require a push
+	appliesToProxy := false
```
Comment (rshriram) [inline pilot/pkg/proxy/envoy/v2/ads.go]: nit for another PR: applies to proxyType reads better.

File: pilot/pkg/proxy/envoy/v2/ads.go [hunk 2]
```diff
@@ -702,25 +706,45 @@ func (s *DiscoveryServer) startPush(req *model.PushRequest) {
 	}
 }
 
-func proxyNeedsPush(con *XdsConnection, targetNamespaces map[string]struct{}) bool {
+func ProxyNeedsPush(proxy *model.Proxy, targetNamespaces map[string]struct{}, configs map[string]struct{}) bool {
```
Comment (rshriram) [inline pilot/pkg/proxy/envoy/v2/ads.go]: I presume this is called before computing the configs?

Comment (howardjohn) [reply pilot/pkg/proxy/envoy/v2/ads.go]: This is called after all init like PushContext generation, but right before XDS config generation

**Pull request merged**